### PR TITLE
stm32U5 has a 12bit  adc4 instance

### PR DESCRIPTION
**Pull request by FRASTM**
Configure the ADC4 peripheral of the stm32U5 serie
-  12 bit resolution
- 19 external analog input channel
-  internal channels including temperature senor and Vbat monitoring channels

Adding an sample to measure the Vbat voltage of the stm32U5  nucleo or disco kit platforms

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -620,6 +647,23 @@ static void adc_stm32_teardown_channels(const struct device *dev, uint8_t channe
...
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
+		if ((__LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_TEMPSENSOR) == channel_id)
+		    && (config->base == ADC1)) {
+			adc_stm32_disable(adc);
+			adc_stm32_unset_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
+		}
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc4), okay)
+		if ((__LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_VBAT_ADC4) == channel_id)
+		   && (config->base == ADC4)) {
```
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:661]: LL_ADC_CHANNEL_TEMPSENSOR_ADC4

**Comment by FRASTM** [comment]
reabse on 64ea7f61ba2bb0a48328fd8aa9998bdd4adbc29c

**Review by gautierg-st — DISMISSED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
The last commit "boards: arm: vref is 3300 on stm32u5 nucleo and disco kit" is valid but there is no reason to apply it only for U5 series.
I'd prefer to exclude it from this PR and make a dedicated PR that applies to all STM32 series (with a dedicated message in the release notes, as it will likely break users).

**Comment by FRASTM** [comment]
> The last commit "boards: arm: vref is 3300 on stm32u5 nucleo and disco kit" is valid but there is no reason to apply it only for U5 series. I'd prefer to exclude it from this PR and make a dedicated PR that applies to all STM32 series (with a dedicated message in the release notes, as it will likely break users).

Moreover in the dts/bindings/adc/st,stm32-adc.yaml,  the property  
```
     vref-mv:
      type: int
      default: 3300
      description: Indicates the reference voltage of the ADC in mV (on the target board).

```
So no need to set the value except for the dts/arm/st/f3/stm32f303.dtsi which has `vref-mv = <3000>;`

**Comment by FRASTM** [comment]
Rebase on bbf03fd9910f4debc991a659d1073abb57e1c843
Removing commit 1a39fca8c0fa046c23c48c375e46979421f40f81 about `vref-mv = <3300>; `of the ADC

File: samples/sensor/stm32_vbat_sensor/boards/b_u585i_iot02a.overlay [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+		compatible = "st,stm32-vbat";
+		ratio = <4>;
+		io-channels = <&adc4 14>;
+		status = "okay";
+	};
+};
+
+&adc4 {
+	has-vbat-channel;
+};
```
Comment (erwango) [inline samples/sensor/stm32_vbat_sensor/boards/b_u585i_iot02a.overlay:20]: I think this should be put directly in stm32u5xxx.dtsi file

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/sensor/stm32_vbat_sensor/boards/b_u585i_iot02a.overlay (hunk 2 above)
Comment (FRASTM) [reply samples/sensor/stm32_vbat_sensor/boards/b_u585i_iot02a.overlay]: done
Note that samples/sensor/stm32_vbat_sensor/boards/nucleo_wb55rg.overlay is not changed in that PR but should also be updated in the same way

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**